File: python/evf/camera/protocol.py
# Copyright (C) 2026 Arun Venkataswamy
#
# This file is part of PushNav.
#
# PushNav is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PushNav is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PushNav. If not, see <https://www.gnu.org/licenses/>.

"""Camera binary protocol — message envelope and type constants.

Per SPEC_PROTOCOL_CAMERA.md §3:
  Every message: type (u32 LE) | length (u32 LE) | payload (length bytes)
"""

import json
import logging
import struct

logger = logging.getLogger(__name__)

# Header: type (u32) + length (u32) = 8 bytes
_HEADER_FMT = "<II"
_HEADER_SIZE = 8

# -- Message types (camera → app) -------------------------------------------
MSG_HELLO = 0x00
MSG_FRAME = 0x01
MSG_CONTROL_INFO = 0x02
MSG_ERROR = 0x03

# -- Message types (app → camera) -------------------------------------------
MSG_SET_CONTROL = 0x11
MSG_GET_CONTROLS = 0x12

_TYPE_NAMES = {
    MSG_HELLO: "HELLO",
    MSG_FRAME: "FRAME",
    MSG_CONTROL_INFO: "CONTROL_INFO",
    MSG_ERROR: "ERROR",
    MSG_SET_CONTROL: "SET_CONTROL",
    MSG_GET_CONTROLS: "GET_CONTROLS",
}


class DisconnectError(Exception):
    """Raised when the remote end closes the connection mid-message."""


def encode_message(msg_type: int, payload: bytes = b"") -> bytes:
    """Encode a protocol message with header + payload."""
    header = struct.pack(_HEADER_FMT, msg_type, len(payload))
    return header + payload


def encode_json_message(msg_type: int, obj: dict) -> bytes:
    """Encode a protocol message with a JSON payload."""
    return encode_message(msg_type, json.dumps(obj).encode("utf-8"))


def _recv_exact(sock, n: int) -> bytes:
    """Read exactly n bytes from sock. Raises DisconnectError on short read."""
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise DisconnectError("Connection closed")
        buf.extend(chunk)
    return bytes(buf)


def read_message(sock) -> tuple[int, bytes]:
    """Read one message from the socket.

    Returns:
        (msg_type, payload) tuple.

    Raises:
        DisconnectError: If the connection closes mid-message.
    """
    header = _recv_exact(sock, _HEADER_SIZE)
    msg_type, length = struct.unpack(_HEADER_FMT, header)
    payload = _recv_exact(sock, length) if length > 0 else b""
    return msg_type, payload


def type_name(msg_type: int) -> str:
    return _TYPE_NAMES.get(msg_type, f"UNKNOWN(0x{msg_type:02x})")

File: camera/mac/Sources/CameraServer/Protocol.swift
// Copyright (C) 2026 Arun Venkataswamy
//
// This file is part of PushNav.
//
// PushNav is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// PushNav is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with PushNav. If not, see <https://www.gnu.org/licenses/>.

// Protocol.swift — Camera binary protocol message envelope and type constants.
//
// Mirrors python/evf/camera/protocol.py.
// Every message: type (u32 LE) | length (u32 LE) | payload (length bytes)

import Foundation

// MARK: - Message Types

enum MessageType: UInt32 {
    // Camera → App
    case hello        = 0x00
    case frame        = 0x01
    case controlInfo  = 0x02
    case error        = 0x03

    // App → Camera
    case setControl   = 0x11
    case getControls  = 0x12

    var name: String {
        switch self {
        case .hello:       return "HELLO"
        case .frame:       return "FRAME"
        case .controlInfo: return "CONTROL_INFO"
        case .error:       return "ERROR"
        case .setControl:  return "SET_CONTROL"
        case .getControls: return "GET_CONTROLS"
        }
    }
}

// MARK: - Header Constants

let protocolHeaderSize = 8  // type (4) + length (4)

// MARK: - Encode

/// Encode a protocol message: 8-byte header (type u32 LE + length u32 LE) + payload.
func encodeMessage(type: MessageType, payload: Data = Data()) -> Data {
    var data = Data(capacity: protocolHeaderSize + payload.count)
    var typeVal = type.rawValue.littleEndian
    var lengthVal = UInt32(payload.count).littleEndian
    data.append(Data(bytes: &typeVal, count: 4))
    data.append(Data(bytes: &lengthVal, count: 4))
    data.append(payload)
    return data
}

/// Encode a protocol message with a JSON-serializable dictionary payload.
func encodeJSONMessage(type: MessageType, object: [String: Any]) -> Data? {
    guard let jsonData = try? JSONSerialization.data(withJSONObject: object) else {
        return nil
    }
    return encodeMessage(type: type, payload: jsonData)
}

// MARK: - Parse Header

/// Parse an 8-byte header, returning (messageType, payloadLength).
/// Returns nil if data is too short or type is unrecognized.
func parseHeader(from data: Data) -> (type: MessageType, length: UInt32)? {
    guard data.count >= protocolHeaderSize else { return nil }
    let typeRaw = data.withUnsafeBytes { $0.load(fromByteOffset: 0, as: UInt32.self) }
    let length = data.withUnsafeBytes { $0.load(fromByteOffset: 4, as: UInt32.self) }
    let typeVal = UInt32(littleEndian: typeRaw)
    let lengthVal = UInt32(littleEndian: length)
    guard let msgType = MessageType(rawValue: typeVal) else { return nil }
    return (msgType, lengthVal)
}
